File: files/9590/lecture5.py
###################################
## Code for POLSCI 9590 Week 4   ##
## Dave Armstrong                ##
## University of Western Ontario ##
## dave.armstrong@uwo.ca         ##
## 2024                          ##
###################################

exec(open('../python_functions.py').read())
import pandas as pd
import numpy as np
from plotnine import ggplot, aes, geom_point, geom_line, stat_summary, theme_bw, theme_classic, labs, ylim, position_dodge, theme, facet_wrap, scale_colour_manual, scale_x_continuous


ces = pd.read_stata("ces19.dta")
sumStats(ces, "leader_lib")
sumStats(ces, "leader_lib", g="educ")


from plotnine import ggplot, aes, stat_summary, theme_classic, labs, ylim
import numpy as np
(ggplot(ces[['educ', 'leader_lib']].dropna(), 
  aes(x='educ', y='leader_lib')) + 
  stat_summary(geom="point", fun_data="mean_cl_normal") + 
  theme_bw() + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)") + 
  ylim(0,100) #<<
).show()



meds = (ces[['educ', 'leader_lib']]
.groupby('educ', observed=True)
.agg(lambda x: x.quantile(.5))
.reset_index()
.rename(columns={"leader_lib": "val"}))
mns = (ces[['educ', 'leader_lib']]
.groupby('educ', observed=True)
.agg(lambda x: np.mean(x))
.reset_index()
.rename(columns={"leader_lib": "val"}))
mns['stat'] = 'Mean'
meds['stat'] = 'Median'
out = meds._append(mns)
out['stat'] = out.stat.astype('category')

(ggplot(out, 
    aes(x="educ", y="val", color="stat", shape="stat")) +  
    geom_point() + 
  theme_bw() + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)", 
       colour="Measure", shape="Measure") + 
  ylim(0,100)).show()



meds = (ces[['educ', 'leader_lib']]
.groupby('educ', observed=True)
.agg(lambda x: x.quantile(.5))
.reset_index()
.rename(columns={"leader_lib": "val"}))
mns = (ces[['educ', 'leader_lib']]
.groupby('educ', observed=True)
.agg(lambda x: np.mean(x))
.reset_index()
.rename(columns={"leader_lib": "val"}))
mns['stat'] = 'Mean'
meds['stat'] = 'Median'
out = meds._append(mns)
out['stat'] = out.stat.astype('category')

(ggplot(out, 
    aes(x="educ", y="val", color="stat", shape="stat")) +  
    geom_point(position = position_dodge(width=.25)) + #<<
  theme_bw() + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)", 
       colour="Measure", shape="Measure") + 
  ylim(0,100)).show()



x = pd.DataFrame(data = {"country": ['A', 'B', 'C'], 
                     "1999": [1,2,3], 
                     "2000": [4,5,6]})
x
x.melt(id_vars="country")


xl = x.melt(id_vars="country")
xl 
xl.pivot(index='country', columns='variable', values='value').reset_index()


cesl = (ces[ces.filter(regex="educ|leader").columns]
  .melt(id_vars = "educ")
)
mns =  (cesl
  .groupby(['educ', 'variable'], observed=True)
  .agg(lambda x: np
  .mean(x))
  .reset_index()
  .rename(columns={"value": "mean"})
  )
meds = (cesl
  .groupby(['educ', 'variable'], observed=True)
  .agg(lambda x: x
  .quantile())
  .reset_index()
  .rename(columns={"value": "median"}))

out = mns.merge(meds).rename(columns = {'variable': 'party'})

outl = pd.DataFrame(out).melt(id_vars = ['educ', 'party'])
caselist = [(outl.party.eq("leader_lib"), "Liberal"), 
            (outl.party.eq("leader_ndp"), "NDP"), 
            (outl.party.eq("leader_con"), "Conservative")]
outl.party = outl.party.case_when(caselist)

(ggplot(outl, aes(x="educ", y="value", 
              colour="party")) + 
  geom_point(position=position_dodge(width=.25)) + 
  theme_bw() + 
  facet_wrap("variable", ncol=2) + 
  scale_colour_manual(values=[ "#003F72", "#d71920", "#F58220"]) + 
  theme(legend_position="top") + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)") + 
  ylim(0,100))



outl['educn'] = outl.educ.cat.codes + 1

(ggplot(outl, aes(x="educn", y="value", 
              colour="party")) + 
  geom_point() + 
  geom_line() + 
  theme_bw() + 
  facet_wrap("variable", ncol=2) + 
  scale_colour_manual(values=[ "#003F72", "#d71920", "#F58220"]) + 
  scale_x_continuous(breaks = [1, 2, 3], labels=["<HS", "HS/College", "College\nGrad"]) + 
  theme(legend_position="top", 
        panel_spacing=.05) + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)") + 
  ylim(0,100)).show()

File: _site/files/9590/lecture5.R
###################################
## Code for POLSCI 9590 Week 4   ##
## Dave Armstrong                ##
## University of Western Ontario ##
## dave.armstrong@uwo.ca         ##
## 2024                          ##
###################################

library(rio)
library(DAMisc)
library(uwo4419)
library(ggplot2)
library(dplyr)
library(tidyr)


ces <- import("ces19.dta")
ces$educ <- factorize(ces$educ)
sumStats(ces, "leader_lib")
sumStats(ces, "leader_lib", byvar="educ")


ces %>% filter(!is.na(educ)) %>% 
ggplot(aes(x=educ, y=leader_lib)) + 
  stat_summary(geom="point", fun=mean) + 
  theme_bw() + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)") + 
  ylim(0,100) #<<
  





ces %>% filter(!is.na(educ)) %>% 
ggplot(aes(x=educ, y=leader_lib)) + 
  stat_summary(aes(shape="Mean", colour="Mean"), #<<
               geom="point", #<<
               fun=mean) + #<<
  stat_summary(aes(shape="Median", colour="Median"), #<<
               geom="point", #<<
               fun=median) + #<<
  theme_bw() + 
  theme(legend.position="top") + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)", 
       colour="Measure", shape="Measure") + #<<
  ylim(0,100) 
  





ces %>% filter(!is.na(educ)) %>% 
ggplot(aes(x=educ, y=leader_lib)) + 
  stat_summary(aes(shape="Mean", colour="Mean"), 
               geom="point", 
               fun=mean, 
               position = position_nudge(x=-.1)) + #<<
  stat_summary(aes(shape="Median", colour="Median"), 
               geom="point",
               fun=median, 
               position = position_nudge(x=.1)) + #<<
  theme_bw() + 
  theme(legend.position="top") + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)", 
       colour="Measure", shape="Measure") + 
  ylim(0,100) 








x <- tibble::tibble(
  country = c("A", "B", "C"), 
  `1999` = 1:3, 
  `2000` = 4:6)
xl <- pivot_longer(x, cols=`1999`:`2000`, 
             names_to="year", 
             values_to="cases")
x
xl





xw <- xl %>% pivot_wider(names_from="year", 
                   values_from="cases")
xl
xw


x <- ces %>% 
  filter(!is.na(educ)) %>% 
  group_by(educ) %>% 
  summarise(across(starts_with("leader"), 
                   list(Mean = ~mean(.x, na.rm=TRUE), 
                        Median = ~median(.x, na.rm=TRUE)))) %>% 
  pivot_longer(-educ, 
               names_pattern="leader_(.*)_(.*)", 
               names_to = c("party", "measure"), 
               values_to="val") %>% 
  mutate(party = factor(party, 
                        levels=c("ndp", "lib", "con"), 
                        labels=c("NDP", "Liberal", "Conservative")))

ggplot(x, aes(x=educ, y=val, 
              colour=party)) + 
  geom_point(position=position_dodge(width=.25)) + 
  theme_bw() + 
  facet_wrap(~measure, ncol=2) + 
  scale_colour_manual(values=c( "#F58220", "#d71920", "#003F72")) + 
  theme(legend.position="top") + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)") + 
  ylim(0,100)





ggplot(x, aes(x=as.numeric(educ), y=val, 
              colour=party)) + 
  geom_point() + 
  geom_line() + 
  theme_bw() + 
  facet_wrap(~measure, ncol=2) + 
  scale_colour_manual(values=c( "#F58220", "#d71920", "#003F72")) + 
  scale_x_continuous(breaks = 1:3, labels=c("<HS", "HS/college", "College\nGrad")) + 
  theme(legend.position="top", 
        panel.spacing=unit(1.5, "lines")) + 
  labs(x="Highest Level of Educational Attainment", 
       y = "Liberal Leader Feeling Thermometer (0-100)") + 
  ylim(0,100)





